src/bluetoothdevice.cpp

```diff
@@ -41,6 +48,8 @@ metric bluetoothdevice::currentResistance() { return Resistance; }
 metric bluetoothdevice::currentCadence() { return Cadence; }
 double bluetoothdevice::currentCrankRevolutions() { return 0; }
 uint16_t bluetoothdevice::lastCrankEventTime() { return 0; }
+
+virtualdevice *bluetoothdevice::VirtualDevice() { return this->virtualDeviceMode==VIRTUAL_DEVICE_MODE::PRIMARY ? this->virtualDevice:nullptr; }
```

Comment:
@drmason789 i guess this `nullptr` could be dangerous. I guess we can simply return the `this->virtualDevice`

what do you think?

by the way I just built and tested the bike and treadmill configuration (also treadmill as a bike, so the alternative one) and everything seems to work (at least with the fake bike and fake treadmill configuration. So it's a good start :)

Reply:
> @drmason789 i guess this `nullptr` could be dangerous. I guess we can simply return the `this->virtualDevice`
> 
> what do you think?

@cagnulein In doing this I am copying the existing behaviour whereby the original VirtualDevice() function only returns one of the virtual device types that could have been set, and nullptr otherwise. 

Considering I'm not in a position to test much of this, I am very reluctant to change behaviour. I aim to match existing behaviour but with more uniformity and ability to maintain from a much smaller number of positions.

If you decide it is alright to return this->virtualDevice, I think we should also consider if the distinction between primary and alternative virtual device mode is required and remove it if we can. At some point there may be tests that pass mocked Zwift data into objects of these classes, and we might test that the expected virtual device is created (I can think how to test this without needing that "firstStateChanged") in which case I would prefer to be checking 1 function.

Reply:
> In doing this I am copying the existing behaviour whereby the original VirtualDevice() function only returns one of the virtual device types that could have been set, and nullptr otherwise.

yes you are right, but since the only calls of this function is from the homeform.cpp i guess it could be safe to return the virtualDevice pointer. of course i will test this

> Considering I'm not in a position to test much of this, I am very reluctant to change behaviour. I aim to match existing behaviour but with more uniformity and ability to maintain from a much smaller number of positions.

Yes i'm highly appreciating your ability to don't change the behavior of the current code giving a better form :)

> If you decide it is alright to return this->virtualDevice, I think we should also consider if the distinction between primary and alternative virtual device mode is required and remove it if we can. At some point there may be tests that pass mocked Zwift data into objects of these classes, and we might test that the expected virtual device is created (I can think how to test this without needing that "firstStateChanged") in which case I would prefer to be checking 1 function.

I actually like the idea to know externally if a virtual device is the primary one or an alternative one, so I would like to leave as you did. 
In the test suite we can just compare if the virtual device is a virtual bike if a device is a bike and the virtual device mode is primary for example 

Ok, i will continue the test, i hope to merge this, this weekend 
Thanks